File: Leetcode/Python/46. Permutations/solution.py
# https://leetcode.com/problems/permutations/

# DFS
from itertools import permutations


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        answer = []
        result = []

        def dfs(numbers: List[int]) -> None:
            if len(result) == len(nums):
                answer.append(result[:])  # If not use copy slice answer list save blank list
                # answer.append(result)
                return
            for idx in range(len(numbers)):
                cur_numbers = numbers[:]  # If not use copy slice del will affect next work
                del cur_numbers[idx]

                result.append(numbers[idx])
                dfs(cur_numbers)
                result.pop()  # Remove calculated number

        dfs(nums)
        return answer


# Use itertools
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        return list(map(list, permutations(nums)))

File: Leetcode/Typescript/46. Permutations/solution.ts
// https://leetcode.com/problems/permutations/description/

function permute(nums: number[]): number[][] {
  const result: number[][] = [];
  function dfs(prev: number[], usedIndex: Set<number>) {
    if (prev.length === nums.length) {
      result.push(prev.slice());
      return;
    }

    for (let idx = 0; idx < nums.length; idx++) {
      if (usedIndex.has(idx)) {
        continue;
      }

      prev.push(nums[idx]);
      usedIndex.add(idx);
      dfs(prev, usedIndex);
      usedIndex.delete(idx);
      prev.pop();
    }
  }
  dfs([], new Set());

  return result;
}
